core-bundle/src/Routing/PageFinder.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Routing;
+
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\PageModel;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\Exception\ExceptionInterface as RoutingExceptionInterface;
+use Symfony\Component\Routing\Matcher\RequestMatcherInterface;
+
+class PageFinder
+{
+    private ContaoFramework $framework;
+    private RequestMatcherInterface $pageRequestMatcher;
+    private RequestMatcherInterface $notFoundRequestMatcher;
+
+    public function __construct(ContaoFramework $framework, RequestMatcherInterface $pageRequestMatcher, RequestMatcherInterface $notFoundRequestMatcher)
+    {
+        $this->framework = $framework;
+        $this->pageRequestMatcher = $pageRequestMatcher;
+        $this->notFoundRequestMatcher = $notFoundRequestMatcher;
+    }
+
+    public function findRootPageByHost(string $hostname, bool $https, string $acceptLanguage): ?PageModel
+    {
+        $request = Request::create(($https ? 'https' : 'http').'://'.$hostname);
+        $request->headers->set('Accept-Language', $acceptLanguage);
+
+        return $this->matchRootPageForRequest($request);
+    }
+
+    public function findRootPageByRequest(Request $request): ?PageModel
+    {
+        $pageModel = $request->attributes->get('pageModel');
+
+        if (!$pageModel instanceof PageModel) {
+            return $this->findRootPageByHost($request->getHost(), $request->isSecure(), $request->headers->get('Accept-Language'));
+        }
+
+        return $this->framework->getAdapter(PageModel::class)->findPublishedById($pageModel->loadDetails()->rootId);
+    }
+
+    public function findFirstPageTypeByRequestHost(Request $request, string $type): ?PageModel
```

Comment:
```suggestion
    public function findFirstPageByTypeRequestHost(Request $request, string $type): ?PageModel
```
“find first page type by…” seems confusing to me :)